File: packages/typescript/src/transport.ts
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";

export type JsonRpcMessage = Record<string, unknown>;

export type MalformedJsonLine = {
  raw: string;
  error: unknown;
};

export type DecodedJsonLines = {
  messages: JsonRpcMessage[];
  malformed: MalformedJsonLine[];
};

export function encodeJsonRpcMessage(message: JsonRpcMessage): string {
  return `${JSON.stringify(message)}\n`;
}

export class JsonLineDecoder {
  #pending = "";

  feed(chunk: string | Buffer): DecodedJsonLines {
    const combined = this.#pending + chunk.toString();
    const parts = combined.split("\n");
    this.#pending = parts.pop() ?? "";

    const messages: JsonRpcMessage[] = [];
    const malformed: MalformedJsonLine[] = [];

    for (const part of parts) {
      const raw = part.endsWith("\r") ? part.slice(0, -1) : part;
      if (raw === "") {
        continue;
      }

      try {
        const parsed = JSON.parse(raw) as unknown;
        if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
          messages.push(parsed as JsonRpcMessage);
        } else {
          malformed.push({ raw, error: new Error("JSON-RPC line must decode to an object") });
        }
      } catch (error) {
        malformed.push({ raw, error });
      }
    }

    return { messages, malformed };
  }

  get pending(): string {
    return this.#pending;
  }
}

export type RoutedMessage =
  | {
      type: "response";
      id: string | number;
      message: JsonRpcMessage;
    }
  | {
      type: "errorResponse";
      id: string | number;
      message: JsonRpcMessage;
    }
  | {
      type: "serverRequest";
      id: string | number;
      message: JsonRpcMessage;
    }
  | {
      type: "orphanResponse";
      id: string | number;
      message: JsonRpcMessage;
    }
  | {
      type: "notification";
      message: JsonRpcMessage;
    }
  | {
      type: "unknown";
      message: JsonRpcMessage;
    };

export class MessageRouter {
  #expectedResponseIds = new Set<string | number>();
  #notifications: JsonRpcMessage[] = [];

  expectResponse(id: string | number): void {
    this.#expectedResponseIds.add(id);
  }

  route(message: JsonRpcMessage): RoutedMessage {
    const id = message.id;

    if (typeof id === "string" || typeof id === "number") {
      if (typeof message.method === "string") {
        return { type: "serverRequest", id, message };
      }

      if (this.#expectedResponseIds.delete(id)) {
        if (message.error && typeof message.error === "object") {
          return { type: "errorResponse", id, message };
        }

        return { type: "response", id, message };
      }

      return { type: "orphanResponse", id, message };
    }

    if (typeof message.method === "string") {
      this.#notifications.push(message);
      return { type: "notification", message };
    }

    return { type: "unknown", message };
  }

  get notifications(): JsonRpcMessage[] {
    return [...this.#notifications];
  }
}

export type AppServerClientOptions = {
  command?: string;
  args?: string[];
  cwd?: string;
  env?: NodeJS.ProcessEnv;
};

export type JsonRpcErrorBody = {
  code: number;
  message: string;
  data?: unknown;
};

export class JsonRpcResponseError extends Error {
  readonly code: number;
  readonly data: unknown;

  constructor(error: JsonRpcErrorBody) {
    super(error.message);
    this.name = "JsonRpcResponseError";
    this.code = error.code;
    this.data = error.data;
  }
}

export class AppServerClosedError extends Error {
  readonly code: number | null;
  readonly signal: NodeJS.Signals | null;

  constructor(code: number | null, signal: NodeJS.Signals | null) {
    super(`app-server closed with code ${code ?? "null"}`);
    this.name = "AppServerClosedError";
    this.code = code;
    this.signal = signal;
  }
}

export type AppServerEvent =
  | {
      type: "notification";
      message: JsonRpcMessage;
    }
  | {
      type: "serverRequest";
      id: string | number;
      message: JsonRpcMessage;
    }
  | {
      type: "malformed";
      raw: string;
    }
  | {
      type: "unknown";
      message: JsonRpcMessage;
    }
  | {
      type: "exit";
      code: number | null;
      signal: NodeJS.Signals | null;
    };

type PendingRequest = {
  resolve: (value: unknown) => void;
  reject: (error: unknown) => void;
};

export class AppServerClient {
  #process: ChildProcessWithoutNullStreams;
  #decoder = new JsonLineDecoder();
  #router = new MessageRouter();
  #pending = new Map<string | number, PendingRequest>();
  #events: AppServerEvent[] = [];
  #eventWaiters: ((event: AppServerEvent) => void)[] = [];
  #requestCounter = 0;
  #closed = false;
  #exitPromise: Promise<void>;
  #resolveExit!: () => void;

  private constructor(process: ChildProcessWithoutNullStreams) {
    this.#process = process;
    this.#exitPromise = new Promise((resolve) => {
      this.#resolveExit = resolve;
    });

    process.stdout.on("data", (chunk: Buffer) => this.#handleStdout(chunk));
    process.on("error", (error) => this.#handleProcessError(error));
    process.on("exit", (code, signal) => this.#handleExit(code, signal));
  }

  static start(options: AppServerClientOptions = {}): AppServerClient {
    const command = options.command ?? "codex";
    const args = options.args ?? ["app-server", "--listen", "stdio://"];
    const child = spawn(command, args, {
      cwd: options.cwd,
      env: options.env,
      stdio: "pipe",
    });

    return new AppServerClient(child);
  }

  request(method: string, params?: unknown): Promise<unknown> {
    const id = `req-${++this.#requestCounter}`;
    this.#router.expectResponse(id);

    const promise = new Promise<unknown>((resolve, reject) => {
      this.#pending.set(id, { resolve, reject });
    });

    this.#write({ id, method, ...(params === undefined ? {} : { params }) });
    return promise;
  }

  async notify(method: string, params?: unknown): Promise<void> {
    this.#write({ method, ...(params === undefined ? {} : { params }) });
  }

  async respond(id: string | number, result: unknown): Promise<void> {
    this.#write({ id, result });
  }

  async respondError(id: string | number, error: JsonRpcErrorBody): Promise<void> {
    this.#write({ id, error });
  }

  nextEvent(): Promise<AppServerEvent> {
    const event = this.#events.shift();
    if (event) {
      return Promise.resolve(event);
    }

    return new Promise((resolve) => {
      this.#eventWaiters.push(resolve);
    });
  }

  events(): AsyncIterable<AppServerEvent> {
    return {
      [Symbol.asyncIterator]: () => ({
        next: async (): Promise<IteratorResult<AppServerEvent>> => {
          if (this.#closed && this.#events.length === 0) {
            return { done: true, value: undefined };
          }

          return { done: false, value: await this.nextEvent() };
        },
      }),
    };
  }

  async close(): Promise<void> {
    if (!this.#closed) {
      this.#process.stdin.end();
      this.#process.kill();
    }

    await this.#exitPromise;
  }

  #write(message: JsonRpcMessage): void {
    if (this.#closed) {
      throw new AppServerClosedError(null, null);
    }

    this.#process.stdin.write(encodeJsonRpcMessage(message));
  }

  #handleStdout(chunk: Buffer): void {
    const decoded = this.#decoder.feed(chunk);
    for (const malformed of decoded.malformed) {
      this.#pushEvent({ type: "malformed", raw: malformed.raw });
    }

    for (const message of decoded.messages) {
      this.#handleMessage(message);
    }
  }

  #handleMessage(message: JsonRpcMessage): void {
    const routed = this.#router.route(message);
    switch (routed.type) {
      case "response": {
        const pending = this.#pending.get(routed.id);
        this.#pending.delete(routed.id);
        pending?.resolve(message.result);
        break;
      }
      case "errorResponse": {
        const pending = this.#pending.get(routed.id);
        this.#pending.delete(routed.id);
        pending?.reject(new JsonRpcResponseError(normalizeJsonRpcError(message.error)));
        break;
      }
      case "serverRequest":
        this.#pushEvent({ type: "serverRequest", id: routed.id, message });
        break;
      case "notification":
        this.#pushEvent({ type: "notification", message });
        break;
      case "unknown":
        this.#pushEvent({ type: "unknown", message });
        break;
      case "orphanResponse":
        this.#pushEvent({ type: "unknown", message });
        break;
    }
  }

  #handleProcessError(error: Error): void {
    for (const pending of this.#pending.values()) {
      pending.reject(error);
    }
    this.#pending.clear();
  }

  #handleExit(code: number | null, signal: NodeJS.Signals | null): void {
    if (this.#closed) {
      return;
    }

    this.#closed = true;
    const error = new AppServerClosedError(code, signal);
    for (const pending of this.#pending.values()) {
      pending.reject(error);
    }
    this.#pending.clear();
    this.#pushEvent({ type: "exit", code, signal });
    this.#resolveExit();
  }

  #pushEvent(event: AppServerEvent): void {
    const waiter = this.#eventWaiters.shift();
    if (waiter) {
      waiter(event);
      return;
    }

    this.#events.push(event);
  }
}

function normalizeJsonRpcError(error: unknown): JsonRpcErrorBody {
  if (error && typeof error === "object") {
    const record = error as Record<string, unknown>;
    return {
      code: typeof record.code === "number" ? record.code : -32000,
      message: typeof record.message === "string" ? record.message : "JSON-RPC error",
      data: record.data,
    };
  }

  return { code: -32000, message: "JSON-RPC error", data: error };
}

File: packages/go/transport.go
package codexsdk

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"slices"
	"strconv"
	"strings"
)

// JSONRPCMessage is a decoded JSON-RPC object.
type JSONRPCMessage map[string]any

// ErrJSONRPCLineNotObject is reported when a newline-delimited JSON value is not an object.
var ErrJSONRPCLineNotObject = errors.New("JSON-RPC line must decode to an object")

// EncodeJSONRPCMessage encodes a JSON-RPC message as compact newline-delimited JSON.
func EncodeJSONRPCMessage(message JSONRPCMessage) ([]byte, error) {
	encoded, err := json.Marshal(message)
	if err != nil {
		return nil, fmt.Errorf("encode JSON-RPC message: %w", err)
	}

	return append(encoded, '\n'), nil
}

// MalformedJSONLine records a JSON line that could not be decoded as a JSON-RPC object.
type MalformedJSONLine struct {
	Raw string
	Err error
}

// DecodedJSONLines contains messages and malformed lines decoded from one feed chunk.
type DecodedJSONLines struct {
	Messages  []JSONRPCMessage
	Malformed []MalformedJSONLine
}

// JSONLineDecoder incrementally decodes newline-delimited JSON-RPC messages.
type JSONLineDecoder struct {
	pending string
}

// NewJSONLineDecoder returns an empty newline-delimited JSON decoder.
func NewJSONLineDecoder() *JSONLineDecoder {
	return &JSONLineDecoder{}
}

// Feed decodes all complete lines from chunk and retains any partial final line.
func (decoder *JSONLineDecoder) Feed(chunk []byte) DecodedJSONLines {
	combined := decoder.pending + string(chunk)
	decoded := DecodedJSONLines{
		Messages:  make([]JSONRPCMessage, 0),
		Malformed: make([]MalformedJSONLine, 0),
	}

	for {
		line, rest, found := strings.Cut(combined, "\n")
		if !found {
			decoder.pending = combined
			return decoded
		}

		combined = rest
		raw, _ := strings.CutSuffix(line, "\r")
		if raw == "" {
			continue
		}

		message, err := decodeJSONRPCLine(raw)
		if err != nil {
			decoded.Malformed = append(decoded.Malformed, MalformedJSONLine{Raw: raw, Err: err})
			continue
		}

		decoded.Messages = append(decoded.Messages, message)
	}
}

// Pending returns the partial line retained for the next feed call.
func (decoder *JSONLineDecoder) Pending() string {
	return decoder.pending
}

func decodeJSONRPCLine(raw string) (JSONRPCMessage, error) {
	var parsed any
	if err := json.Unmarshal([]byte(raw), &parsed); err != nil {
		return nil, fmt.Errorf("decode JSON-RPC line: %w", err)
	}

	message, ok := parsed.(map[string]any)
	if !ok {
		return nil, ErrJSONRPCLineNotObject
	}

	return JSONRPCMessage(message), nil
}

// RouteType identifies how a decoded JSON-RPC message was routed.
type RouteType string

const (
	// RouteResponse is an expected response for a tracked request ID.
	RouteResponse RouteType = "response"
	// RouteErrorResponse is an expected JSON-RPC error response for a tracked request ID.
	RouteErrorResponse RouteType = "errorResponse"
	// RouteServerRequest is a request initiated by the server.
	RouteServerRequest RouteType = "serverRequest"
	// RouteOrphanResponse is a response for an untracked request ID.
	RouteOrphanResponse RouteType = "orphanResponse"
	// RouteNotification is a JSON-RPC notification.
	RouteNotification RouteType = "notification"
	// RouteUnknown is an object that is neither a response nor a notification.
	RouteUnknown RouteType = "unknown"
)

// RoutedMessage is the result of routing a decoded JSON-RPC message.
type RoutedMessage struct {
	Type    RouteType
	ID      any
	Message JSONRPCMessage
}

// MessageRouter routes decoded JSON-RPC objects into responses, notifications, and unknown messages.
type MessageRouter struct {
	expectedResponseIDs map[responseID]struct{}
	notifications       []JSONRPCMessage
}

// NewMessageRouter returns an empty message router.
func NewMessageRouter() *MessageRouter {
	return &MessageRouter{expectedResponseIDs: make(map[responseID]struct{})}
}

// ExpectResponse tracks requestID as an expected response.
func (router *MessageRouter) ExpectResponse(requestID any) {
	key, ok := normalizeResponseID(requestID)
	if !ok {
		return
	}

	router.expectedResponseIDs[key] = struct{}{}
}

// Route routes a decoded JSON-RPC object.
func (router *MessageRouter) Route(message JSONRPCMessage) RoutedMessage {
	if id, ok := message["id"]; ok {
		if key, ok := normalizeResponseID(id); ok {
			if _, ok := message["method"].(string); ok {
				return RoutedMessage{Type: RouteServerRequest, ID: id, Message: message}
			}

			if _, expected := router.expectedResponseIDs[key]; expected {
				delete(router.expectedResponseIDs, key)
				if _, ok := message["error"].(map[string]any); ok {
					return RoutedMessage{Type: RouteErrorResponse, ID: id, Message: message}
				}

				return RoutedMessage{Type: RouteResponse, ID: id, Message: message}
			}

			return RoutedMessage{Type: RouteOrphanResponse, ID: id, Message: message}
		}
	}

	if method, ok := message["method"].(string); ok && method != "" {
		router.notifications = append(router.notifications, message)
		return RoutedMessage{Type: RouteNotification, Message: message}
	}

	return RoutedMessage{Type: RouteUnknown, Message: message}
}

// Notifications returns the notifications routed so far.
func (router *MessageRouter) Notifications() []JSONRPCMessage {
	return slices.Clone(router.notifications)
}

type responseID struct {
	kind  string
	value string
}

func normalizeResponseID(id any) (responseID, bool) {
	switch value := id.(type) {
	case string:
		return responseID{kind: "string", value: value}, true
	case int:
		return responseID{kind: "number", value: strconv.FormatInt(int64(value), 10)}, true
	case int64:
		return responseID{kind: "number", value: strconv.FormatInt(value, 10)}, true
	case float64:
		if math.IsNaN(value) || math.IsInf(value, 0) {
			return responseID{}, false
		}

		return responseID{kind: "number", value: strconv.FormatFloat(value, 'f', -1, 64)}, true
	case json.Number:
		return responseID{kind: "number", value: value.String()}, true
	default:
		return responseID{}, false
	}
}
